File: w06/P2.py
class Item:
  def __init__(self, w, v):
    self.weight = w
    self.value = v

class Knapsack:
  def __init__(self, items, cap):
    self.items = items
    self.bestSubset = [False] * len(self.items)
    self.maxValue = 0
    self.capacity = cap

  def start(self):
    self.subset([False] * len(self.items), 0)

  def subset(self, selected, idx):
    if (idx == len(self.items)):
      self.process(selected)
      return

    # Not selected
    selected[idx] = False
    self.subset(selected, idx + 1)

    # Selected
    selected[idx] = True
    self.subset(selected, idx + 1)

  def process(self, selected):
    w = 0
    v = 0
    for i in range(len(selected)):
      if (selected[i]):
        w += self.items[i].weight
        v += self.items[i].value
        if (w > self.capacity):
          return

    if (v > self.maxValue):
      self.maxValue = v
      self.bestSubset = selected.copy()

  def displayBest(self):
    res = "Best subset: "
    for i in range(len(self.bestSubset)):
      if (self.bestSubset[i]):
        res = res + str(i) + ", "
    res = res + "with total value: " + str(self.maxValue)
    print(res)


items = []
items.append(Item(7, 42))
items.append(Item(3, 12))
items.append(Item(4, 40))
items.append(Item(5, 25))

knapsack = Knapsack(items, 10)
knapsack.start()
knapsack.displayBest()

File: w06/P2.java
package w06;

public class P2 {
  Item[] items;
  boolean[] bestSubset;
  int maxValue;
  int capacity;

  public P2(Item[] i, int c) {
    items = i;
    bestSubset = new boolean[i.length];
    capacity = c;
    maxValue= 0;
  }

  public void start() {
    subset(new boolean[items.length], 0);
  }

  void subset(boolean[] selected, int idx) {
    if (idx == items.length) {
      process(selected);
      return;
    }

    // Not selected
    selected[idx] = false;
    subset(selected, idx + 1);

    // Selected
    selected[idx] = true;
    subset(selected, idx + 1);
  }

  void process(boolean[] selected) {
    int w = 0, v = 0;
    for (int i = 0; i < selected.length; i++) {
      if (selected[i]) {
        w += items[i].weight;
        v += items[i].value;
        if (w > capacity) {
          return;
        }
      }
    }
    if (v > maxValue) {
      maxValue = v;
      bestSubset = selected.clone();
    }
  }

  void displayBest() {
    StringBuilder res = new StringBuilder("Best subset: ");
    for (int i = 0; i < bestSubset.length; i++) {
      if (bestSubset[i]) {
        res.append(i + ", ");
      }
    }
    res.append("with total value: " + maxValue);
    System.out.println(res);
  }

  public static void main(String[] args) {
    Item[] items = new Item[] {
      new Item(7, 42),
      new Item(3, 12),
      new Item(4, 40),
      new Item(5, 25)
    };
    P2 knapsack = new P2(items, 10);
    knapsack.start();
    knapsack.displayBest();
  }
}

class Item {
  int weight;
  int value;

  public Item(int w, int v) {
    weight = w;
    value = v;
  }
}